### PR TITLE
Roi

### PR DESCRIPTION
**Pull request by bnmajor**
Depends on Kitware/itk-vtk-viewer#724 and #702

### CODE REVIEW

File: itkwidgets/viewer.py
Comment (thewtex) [inline itkwidgets/viewer.py:246]: nit: for the Python argument use snake case, `croppingPlanes` -> `cropping_planes`

File: itkwidgets/viewer.py
Comment (thewtex) [inline itkwidgets/viewer.py:357]: `l` -> `L`

File: itkwidgets/viewer.py
Comment (thewtex) [inline itkwidgets/viewer.py:371]: very cool!!

File: itkwidgets/viewer.py
Comment (bnmajor) [reply itkwidgets/viewer.py:246]: :100:

**Comment by bnmajor** [comment]
@thewtex  @PaulHax  A new example notebook has been added but if there is better data to use (maybe a large dataset that is typically stuck to a lower scale?) I'd be happy to update with a different example. This data was more or less selected at random.

**Comment by thewtex** [comment]
<p>@bnmajor awesome job!</p><p> </p><p>A few requests inline.</p><p> </p><p>This is a good dataset, but another to consider is:</p><p> </p><p><a href="https://data.kitware.com/#item/63476e0911dab81428208e27" target="_blank">https://data.kitware.com/#item/63476e0911dab81428208e27</a></p><p> </p><p>for more scales / variety.</p>

**Comment by thewtex** [comment]
<p>@bnmajor awesome job!</p><p> </p><p>A few requests inline.</p><p> </p><p>This is a good dataset, but another to consider is:</p><p> </p><p><a href="https://data.kitware.com/#item/63476e0911dab81428208e27" target="_blank">https://data.kitware.com/#item/63476e0911dab81428208e27</a></p><p> </p><p>for more scales / variety.</p>

File: examples/integrations/itk/SelectROI.ipynb
Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <p>This gif is wicked cool to have embedded in the notebook!! How did you do that? :-D</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    loaded_scale
</pre><p>Let's make a note, provide a reminder on how the scales, are defined, i.e. "Zero is the highest resolution scale."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    roi_region
</pre><p>Let's make a note / provide a reminder that these are in physical, world coordinates.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    # Request the highest level explicitly in case that is not what is loaded.
</pre><p>To make the connection clearer, let's use <code>loaded_scale</code> here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    roi_slice
</pre><p>Let's add another cell for illustration that calls <code>viewer.get_roi_slice()</code> (no argument) that demonstrates we get the same result since that is the current scale. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (bnmajor) [reply examples/integrations/itk/SelectROI.ipynb:1]: It's just an html image tag in a markdown cell! Updated to use the full URL instead of the relative path though so it may appear broken in the review notebook now until these changes are merged in.

Comment (bnmajor) [reply examples/integrations/itk/SelectROI.ipynb:1]: Sounds good, updated :+1:

Comment (bnmajor) [reply examples/integrations/itk/SelectROI.ipynb:1]: Got it!

Comment (bnmajor) [reply examples/integrations/itk/SelectROI.ipynb:1]: Sounds good!

**Comment by thewtex** [comment]
<p>Awesome!</p>

File: examples/integrations/itk/SelectROI.ipynb
Comment (thewtex) [inline examples/integrations/itk/SelectROI.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    print(roi_slices, default_roi_slices)
</pre><p>We could make these two print statements so they are easier to compare.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/InsightSoftwareConsortium/itkwidgets/pull/703/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Review by thewtex — APPROVED** [review]

**Comment by bnmajor** [comment]
@thewtex Two last items:
- I did try the other dataset that you suggested but I did not use it because it appears to be missing spacing information?
missing_spacing
- When we grab the ROI for lower scales it seems that we lose the metadata or something... The resulting image is skewed (this does not happen at scale 0):
skewed_lower_scales

File: examples/integrations/itk/SelectROI.ipynb
Comment (bnmajor) [reply examples/integrations/itk/SelectROI.ipynb:1]: Got it!

**Comment by thewtex** [comment]
@bnmajor what is your `ngff_zarr.__version__`?

**Comment by thewtex** [comment]
Checking this branch out locally, I could not reproduce with ngff-Zarr 0.6.0.

**Comment by bnmajor** [comment]
> Checking this branch out locally, I could not reproduce with ngff-Zarr 0.6.0.

I was on `0.4.7`... Sorry about that!

**Comment by thewtex** [comment]
I seem to recall an older version of itk-vtk-viewer causing this behavior -- is it possible that an old version is used locally?

**Comment by thewtex** [comment]
I tried again with a fresh conda environment and repository checkout, the spacing was correct and consistent across scales.

**Comment by bnmajor** [comment]
I'm not using a local instance of the itk-vtk-viewer and we're currently using the latest release...
skewed_lower_scale

**Comment by thewtex** [comment]
Oh, I was just looking and `viewer1` -- I do see that with `viewer2` :-)

**Comment by thewtex** [comment]
> roi = multiscales.images[loaded_scale].data[roi_slices]

This is extracting only the array and discarding the metadata.

**Comment by thewtex** [comment]
image

with:

```
# Create a new viewer using only the data in the ROI determined above
loaded_image = multiscales.images[loaded_scale]
roi_data = loaded_image.data[roi_slices]

from ngff_zarr import to_ngff_image

roi_image = to_ngff_image(roi_data, dims=loaded_image.dims, scale=loaded_image.scale, translation={'x': roi_region[0][0], 'y': roi_region[0][1], 'z': roi_region[0][2]})

from ngff_zarr import to_ngff_image
```

0. update the notebook according to the above.
1. could you please update get_roi_region to output a list of dictionaries, 'x', 'y', etc.

?

**Comment by bnmajor** [comment]
> 0. update the notebook according to the above.
> 1. could you please update get_roi_region to output a list of dictionaries, 'x', 'y', etc.
> 
> ?

Updates have been pushed, thank you

**Comment by thewtex** [comment]
Sorry, I was not too clear -- I will push a patch with a minor tweak.

**Comment by bnmajor** [comment]
> Sorry, I was not too clear -- I will push a patch with a minor tweak.

I see, sorry about my misunderstanding! Thanks for the fix!

**Comment by bnmajor** [comment]
@thewtex This is good to go from my end if everything is looking good for you as well

**Pull request merged**
